addons/l10n_in_ewaybill/models/l10n_in_ewaybill.py

```diff
@@ -0,0 +1,654 @@
+                ),
+                # shipping details
+                **prepare_details(
+                    key_paired_function={
+                        "Addr1": lambda p: p.street and p.street[:120] or "",
+                        "Addr2": lambda p: p.street2 and p.street2[:120] or "",
+                        "Place": lambda p: p.city and p.city[:50] or "",
+                        "Pincode": lambda p: int(p.zip) if p.country_id.code == "IN" else 999999,
+                    }.items(),
+                    partner_detail={'from': self.partner_ship_from_id, 'to': self.partner_ship_to_id}.items()
+                ),
+                "actToStateCode": self._get_partner_state_code(self.partner_ship_to_id),
+                "actFromStateCode": self._get_partner_state_code(self.partner_ship_from_id),
+        }
+        return ewaybill_json
+
+    def _prepare_ewaybill_transportation_json_payload(self):
+        # only pass transporter details when value is exist
+        return dict(
+            filter(lambda kv: kv[1], {
+                "transporterId": self.transporter_id.vat,
+                "transporterName": self.transporter_id.name,
+                "transMode": self.mode,
+                "transDocNo": self.transportation_doc_no,
+                "transDocDate": self.transportation_doc_date and self.transportation_doc_date.strftime("%d/%m/%Y"),
+                "vehicleNo": self.vehicle_no,
+                "vehicleType": self.vehicle_type,
+            }.items())
+        )
+
+    def _prepare_ewaybill_tax_details_json_payload(self):
+        round_value = self.env['account.move']._l10n_in_round_value
+        tax_details = self.account_move_id._l10n_in_prepare_tax_details()
+        tax_details_by_code = self.env['account.move']._get_l10n_in_tax_details_by_line_code(tax_details.get("tax_details", {}))
+        invoice_line_tax_details = tax_details.get("tax_details_per_record")
+        return {
+            "itemList": list(starmap(self._get_l10n_in_ewaybill_line_details, invoice_line_tax_details.items())),
+            "totalValue": round_value(tax_details.get("base_amount", 0.00)),
+            **{
+                f'{tax_type}Value': round_value(tax_details_by_code.get(f'{tax_type}_amount', 0.00))
+                for tax_type in ['cgst', 'sgst', 'igst', 'cess']
+            },
+            "cessNonAdvolValue": round_value(tax_details_by_code.get("cess_non_advol_amount", 0.00)),
+            "otherValue": round_value(tax_details_by_code.get("other_amount", 0.00)),
+            "totInvValue": round_value(tax_details.get("base_amount", 0.00) + tax_details.get("tax_amount", 0.00)),
+        }
+
+    def _ewaybill_generate_direct_json(self):
+        return {
+            **self._prepare_ewaybill_base_json_payload(),
+            **self._prepare_ewaybill_transportation_json_payload(),
+            **self._prepare_ewaybill_tax_details_json_payload(),
+        }
+
+    def _set_data_from_attachment(self):
```

Comment:
Could use a docstring so we know it is used for upgrade.

Reply:
Done.